scripts/bench_literals.go

```diff
@@ -0,0 +1,164 @@
+)
+
+func generateRunSrc() string {
+	var sb strings.Builder
+
+	sb.WriteString(`
+var alwaysFalseFlag = false
+
+func noop(i interface{}) {
+	if alwaysFalseFlag {
+		println(i)
+	}	
+}
+
+func Run() {
+`)
+
+	dataStr := strings.Repeat("X", maxDataLen)
+	dataBytes := make([]string, maxDataLen)
+	for i := 0; i < len(dataBytes); i++ {
+		dataBytes[i] = strconv.Itoa(i)
+	}
+
+	for dataLen := minDataLen; dataLen <= maxDataLen; dataLen += stepDataLen {
+		for y := 0; y < dataCountPerLen; y++ {
+			fmt.Fprintf(&sb, "\tnoop(%q)\n", dataStr[:dataLen])
+			fmt.Fprintf(&sb, "\tnoop([]byte{%s})\n", strings.Join(dataBytes[:dataLen], ", "))
+		}
+	}
+
+	sb.WriteString("}\n")
+	return sb.String()
+}
+
+func buildTestGarble(tdir string) string {
+	garbleBin := filepath.Join(tdir, "garble")
+	if runtime.GOOS == "windows" {
+		garbleBin += ".exe"
+	}
+
+	output, err := exec.Command("go", "build", "-tags", "garble_testing", "-o="+garbleBin).CombinedOutput()
+	if err != nil {
+		log.Fatalf("garble build failed: %v\n%s", err, string(output))
+	}
+
+	return garbleBin
+}
+
+func writeDateFile(tdir, obfName, src string) error {
+	pkgName := packagePrefix + obfName
+
+	dir := path.Join(tdir, pkgName)
+	if err := os.MkdirAll(dir, 0o777); err != nil {
+		return err
+	}
+
+	f, err := os.Create(path.Join(dir, "data.go"))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	fmt.Fprintf(f, "package %s\n\n", pkgName)
+	f.WriteString(src)
+	return nil
+}
+
+func writeTestFile(dir, obfName string) error {
+	f, err := os.Create(path.Join(dir, obfName+"_test.go"))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	f.WriteString(`package main
+import "testing"
+`)
+	pkgName := packagePrefix + obfName
+	fmt.Fprintf(f, "import %q\n", moduleName+"/"+pkgName)
+	fmt.Fprintf(f, `func Benchmark%s(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		%s.Run()		
+	}
+}
+`, strings.ToUpper(obfName[:1])+obfName[1:], pkgName)
+	return nil
+}
+
+func main() {
+	tdir, err := os.MkdirTemp("", "literals-bench*")
+	if err != nil {
+		log.Fatalf("create temp directory failed: %v", err)
+	}
+	defer os.RemoveAll(tdir)
+
+	if err := os.WriteFile(path.Join(tdir, "go.mod"), []byte("module "+moduleName), 0o777); err != nil {
+		log.Fatalf("write go.mod failed: %v", err)
+	}
+
+	runSrc := generateRunSrc()
+	writeTest := func(name string) {
+		if err := writeDateFile(tdir, name, runSrc); err != nil {
+			log.Fatalf("write data for %s failed: %v", name, err)
+		}
+		if err := writeTestFile(tdir, name); err != nil {
+			log.Fatalf("write test for %s failed: %v", name, err)
+		}
+	}
+
+	var packageToObfuscatorIndex []string
+	for i, obf := range literals.Obfuscators {
+		obfName := reflect.TypeOf(obf).Name()
+		writeTest(obfName)
+		packageToObfuscatorIndex = append(packageToObfuscatorIndex, fmt.Sprintf(packagePrefix+"%s=%d", obfName, i))
+	}
+	writeTest("all")
+
+	garbleBin := buildTestGarble(tdir)
+	args := append([]string{"-seed", garbleSeed, "-literals", "test", "-bench"}, os.Args[1:]...)
```

Comment:
clever to use `garble test -bench` :)

just so we're on the same page, this will only benchmark running the function containing the code with the literals. we're not benchmarking building obfuscated binaries with `-literals`, nor are we benchmarking the `init` cost of a program obfuscated with `-literals`. all that sounds fine with me, but we should clarify what the benchmark does and doesn't cover, perhaps at the top of the file in a comment.